File: chap14/fig14-1.R
library(survival)
library(ggplot2)

d <- read.csv('input/data-surv.csv')
d_plot <- d[1:15, ]
d_plot$Join <- as.Date(d_plot$Join)
d_plot$Leave <- as.Date(d_plot$Leave)

p <- ggplot() +
  theme_bw(base_size=16) +
  geom_segment(data=d_plot, aes(x=Join, y=PersonID, xend=Leave, yend=PersonID), size=1) +
  geom_point(data=d_plot, aes(x=Join, y=PersonID), size=2, shape=16) +
  geom_point(data=subset(d_plot, Cens == 0), aes(x=Leave, y=PersonID), size=2, shape=16) +
  geom_point(data=subset(d_plot, Cens == 1), aes(x=Leave, y=PersonID), size=3, shape=4) +
  scale_y_reverse() +
  xlab('Date')
ggsave(p, file='output/fig14-1-left.png', dpi=300, width=3, height=3)


p <- ggplot() +
  theme_bw(base_size=16) +
  geom_histogram(data=d, aes(x=Time), binwidth=3, color='black', fill='white')
ggsave(p, file='output/fig14-1-middle.png', dpi=300, width=3, height=3)


res <- survfit(Surv(Time, 1-Cens) ~ 1, data=d)
png('output/fig14-1-right.png', pointsize=48, width=1300, height=1150)
plot(res, mark.time=TRUE, lwd=2, xlab='Time (Months)', ylab='Survival')
dev.off()

File: chap14/fig14-1.py
import pandas
from lifelines import KaplanMeierFitter
from plotnine import ggplot, aes, theme_bw, theme, element_text, geom_segment, geom_point, scale_x_datetime, scale_y_reverse, xlab, geom_histogram
from mizani.formatters import date_format
from matplotlib import pyplot as plt

d = pandas.read_csv('input/data-surv.csv', parse_dates=['Join', 'Leave'])
d_plot = d.iloc[0:15,:]

p = (ggplot()
    + theme_bw(base_size=16)
    + theme(axis_text_x=element_text(angle=40, vjust=1, hjust=1))
    + geom_segment(d_plot, aes('Join', 'PersonID', xend='Leave', yend='PersonID'), size=1)
    + geom_point(d_plot, aes('Join', 'PersonID'), size=2, shape='o')
    + geom_point(d_plot[d_plot.Cens == 0].reset_index(), aes('Leave', 'PersonID'), size=2, shape='o')
    + geom_point(d_plot[d_plot.Cens == 1].reset_index(), aes('Leave', 'PersonID'), size=3, shape='x')
    + scale_y_reverse()
    + scale_x_datetime(labels = date_format("%Y"))
    + xlab('Date')
)
p.save(filename='output/fig14-1-left.py.png', dpi=300, width=3, height=3)


p = (ggplot()
    + theme_bw(base_size=16)
    + geom_histogram(d, aes('Time'), binwidth=3, color='black', fill='white')
)
p.save(filename='output/fig14-1-middle.py.png', dpi=300, width=3, height=3)


fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)
kmf = KaplanMeierFitter()
kmf.fit(d.Time, event_observed=d.Cens-1)
ax = kmf.plot()
fig.savefig('output/fig14-1-right.py.png', dpi=200)
